subsystems/mvdm/ntvdm/bios/vidbios.c

```diff
@@ -2942,6 +2942,295 @@ static VOID VidBiosPrintCharacter(CHAR Character, BYTE Attribute, BOOLEAN UseAtt
     VidBiosSetCursorPosition(Row, Column, Page);
 }
 
+static VOID VidBiosSetPixel(WORD X, WORD Y, BYTE Page, BYTE Value)
+{
+    SHORT ScreenWidth = VgaGetDisplayResolution().X;
+
+    if (IS_TEXT_MODE(Bda->VideoMode))
+    {
+        DPRINT1("Note, trying to modify pixel (%d,%d) @ page %d to %d in text mode %X - returning 0.\n", X, Y, Page, Value, Bda->VideoMode);
+        return;
+    }
+
+    switch (Bda->VideoMode)
+    {
+        case 0x04: /* CGA 320x200 4c */
+        case 0x05: /* CGA 320x200 4c */
+        case 0x06: /* CGA 640x200 2c  */
+        {
+            ULONG Address;
+            BYTE MemoryValue;
+            BOOL Xor;
+            BYTE MaxColor;
+            BYTE ValueShift;
+            BYTE PixelsPerByte;
+
+            if (Bda->VideoMode == 0x06)
+            {
+                MaxColor = 1;
+                ValueShift = 7 - X % 8;
+                PixelsPerByte = 8;
+            }
+            else
+            {
+                MaxColor = 3;
+                ValueShift = (3 - X % 4) * 2;
+                PixelsPerByte = 4;
+            }
+
+            Xor = (Value & 0x80) != 0;
+            Value &= MaxColor;
+            Address = TO_LINEAR((Y & 1) ? CGA_ODD_VIDEO_SEG : CGA_EVEN_VIDEO_SEG,
+                                Page * Bda->VideoPageSize + ((Y / 2) * ScreenWidth + X) / PixelsPerByte);
+
+            EmulatorReadMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            if (Xor)
+            {
+                MemoryValue ^= Value << ValueShift;
+            }
+            else
+            {
+                MemoryValue &= ~(MaxColor << ValueShift);
+                MemoryValue |= Value << ValueShift;
+            }
+            EmulatorWriteMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            break;
+        }
+
+        case 0x0D: /* EGA 320x200 16c */
+        case 0x0E: /* EGA 640x200 16c */
+        case 0x10: /* EGA 640x350 16c */
+        case 0x12: /* VGA 640x480 16c */
+        {
+            ULONG Address;
+            BOOL Xor;
+            BYTE OldMode;
+            BYTE OldRotation;
+            BYTE OldRead;
+            BYTE OldMask;
+            BYTE CurrentPlaneMask;
+            BYTE CurrentPlane;
+            BYTE Shift = 7 - X % 8;
+
+            Xor = (Value & 0x80) != 0;
+            Value &= 15;
```

Comment:
```suggestion
            Value &= 0x0F;
```

Reply:
My thought while writing this was that & 15 better represents "hey I'm trimming the value to 16 colors" than just & 0x0F